<a href="https://colab.research.google.com/github/Thenjiwe2603/Viewership-Data-ECD/blob/main/Viewership_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Importing Libraries**

In [58]:
# Brings in the pandas library,
# giving it alias pd and excellent for working with structured data
import pandas as pd
# Giving it alias nd and excellent for working with numerical calculations
import numpy as nd
import datetime
import time

**Data Ingestion**

In [59]:
# Location of the dataset
data_path = ("/content/drive/MyDrive/Viewership Analysis .xlsx - Data.csv")

In [60]:
# Reading the data
Viewers_Analysis=pd.read_csv(data_path)
df=pd.read_csv(data_path)


In [61]:
# Displaying data
display(Viewers_Analysis)

DateID                    CustomerID  TotalTimeWatched  Platform  \
0       20201101  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
1       20201101  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
2       20201101  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
3       20201101  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
4       20201101  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
...          ...                           ...               ...       ...   
118529  20210416  DGXM5TJ18W1LANOLKJJN5MFT000G               600  Leanback   
118530  20210416  DMMVDBZFDT1F7WL4BGWQ5KOJ000G              5194       Web   
118531  20210416  1JKBIAY7MEUH6BWX0SCU93GJ000I             47660       Web   
118532  20210416  1ADX18R3B33REZ2T0NK911E4002W              1335  Leanback   
118533  20210416  BO0REPHA6FP7C15B0K9N1GVO0050               360       iOS   

       PlayEventType                           VideoTitle  
0             LiveTV            F1 '20: Emilia Romagna GP  
1             LiveTV            F1 '20: Emilia Romagna GP  
2              Other                      Chasing The Sun  
3              Other                      Chasing The Sun  
4             LiveTV                   Sonic The Hedgehog  
...              ...                                  ...  
118529        LiveTV                              New Day  
118530        LiveTV  Buying And Selling With Property...  
118531        LiveTV                             Expresso  
118532        LiveTV                              Unknown  
118533        LiveTV                               Imbewu  

[118534 rows x 6 columns]

**3. Exploratory Data Analysis (EDA)**


In [62]:
# Returns a tuple in a format of number of rows and columns of the dataframe
df.shape

(118534, 6)

**Observations**

*  The dataset has 118534 rows
*  The dataset has 6 columns



In [63]:
# Showing the first (10) rows of dataframe
df.head(10)

DateID                    CustomerID  TotalTimeWatched  Platform  \
0  20201101  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
1  20201101  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
2  20201101  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
3  20201101  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
4  20201101  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
5  20201101  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
6  20201101  6PMV67PLJ2S47S68J0Y30XFK003C               120       iOS   
7  20201101  6PMV67PLJ2S47S68J0Y30XFK003C               120       iOS   
8  20201101  6PMV67PLJ2S47S68J0Y30XFK003C                60       iOS   
9  20201101  6PMV67PLJ2S47S68J0Y30XFK003C               600       iOS   

  PlayEventType                   VideoTitle  
0        LiveTV    F1 '20: Emilia Romagna GP  
1        LiveTV    F1 '20: Emilia Romagna GP  
2         Other              Chasing The Sun  
3         Other              Chasing The Sun  
4        LiveTV           Sonic The Hedgehog  
5        LiveTV                    Frozen II  
6        LiveTV  PL 20/21: Man Utd v Arsenal  
7        LiveTV  PL 20/21: Man Utd v Arsenal  
8        LiveTV    F1 '20: Emilia Romagna GP  
9        LiveTV    F1 '20: Emilia Romagna GP

**Observations**
* Function gives the first rows of a dataframe,
  in this example have selected first 10 rows



In [64]:
# Showing the last (8) rows of dataframe
Viewers_Analysis.tail(8)

DateID                    CustomerID  TotalTimeWatched  Platform  \
118526  20210416  HQLC2AR9CR5KG6F9CKXXJLHY003D               238       iOS   
118527  20210416  HQLC2AR9CR5KG6F9CKXXJLHY003D              2641  Leanback   
118528  20210416  M1NMIW7FFTPM0V2SMBJAAT8P005I              2460  Leanback   
118529  20210416  DGXM5TJ18W1LANOLKJJN5MFT000G               600  Leanback   
118530  20210416  DMMVDBZFDT1F7WL4BGWQ5KOJ000G              5194       Web   
118531  20210416  1JKBIAY7MEUH6BWX0SCU93GJ000I             47660       Web   
118532  20210416  1ADX18R3B33REZ2T0NK911E4002W              1335  Leanback   
118533  20210416  BO0REPHA6FP7C15B0K9N1GVO0050               360       iOS   

       PlayEventType                           VideoTitle  
118526        LiveTV                          PL Build Up  
118527        LiveTV        PL 20/21: Everton v Tottenham  
118528        LiveTV                              Unknown  
118529        LiveTV                              New Day  
118530        LiveTV  Buying And Selling With Property...  
118531        LiveTV                             Expresso  
118532        LiveTV                              Unknown  
118533        LiveTV                               Imbewu

**Observations**

* This function gives the last rows of a dataframe,
  in this example have selected first 8 rows



In [65]:
# Gives statistical numerical columns in your dataset
df.describe()

DateID  TotalTimeWatched
count  1.185340e+05     118534.000000
mean   2.020945e+07       2046.869860
std    2.583117e+03       3735.512071
min    2.020110e+07          1.000000
25%    2.021012e+07        240.000000
50%    2.021022e+07       1020.000000
75%    2.021032e+07       2400.000000
max    2.021042e+07      88500.000000

**Observations**

*  **count**  : The number of non-null values
*  **mean**   : The average value.
*   **max**    : The maximum value etc



In [66]:
# Gives column names with the datatype of each column
df.dtypes

DateID               int64
CustomerID          object
TotalTimeWatched     int64
Platform            object
PlayEventType       object
VideoTitle          object
dtype: object

**Observations**

*   DateID column is an integer datatype
*   CustomerID column is object
*   TotalTimeWatched column is integer
*   Platform is an object
*   PlayEventType is an object
*   VideoTitle is object







In [67]:
df_cleaned=df.drop_duplicates() # Dropping duplicates
print(df_cleaned)

          DateID                    CustomerID  TotalTimeWatched  Platform  \
0       20201101  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
2       20201101  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
4       20201101  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
5       20201101  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
6       20201101  6PMV67PLJ2S47S68J0Y30XFK003C               120       iOS   
...          ...                           ...               ...       ...   
118529  20210416  DGXM5TJ18W1LANOLKJJN5MFT000G               600  Leanback   
118530  20210416  DMMVDBZFDT1F7WL4BGWQ5KOJ000G              5194       Web   
118531  20210416  1JKBIAY7MEUH6BWX0SCU93GJ000I             47660       Web   
118532  20210416  1ADX18R3B33REZ2T0NK911E4002W              1335  Leanback   
118533  20210416  BO0REPHA6FP7C15B0K9N1GVO0050               360       iOS   

       PlayEventType                           VideoTitle  
0  

In [68]:
df_cleaned.duplicated().sum() #After dropping duplicates, check did it drop duplicates

np.int64(0)

In [69]:
df['PlayEventType'].unique() # Finds all distinct (non-duplicate) values that appear in that column
print(df)

          DateID                    CustomerID  TotalTimeWatched  Platform  \
0       20201101  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
1       20201101  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
2       20201101  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
3       20201101  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
4       20201101  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
...          ...                           ...               ...       ...   
118529  20210416  DGXM5TJ18W1LANOLKJJN5MFT000G               600  Leanback   
118530  20210416  DMMVDBZFDT1F7WL4BGWQ5KOJ000G              5194       Web   
118531  20210416  1JKBIAY7MEUH6BWX0SCU93GJ000I             47660       Web   
118532  20210416  1ADX18R3B33REZ2T0NK911E4002W              1335  Leanback   
118533  20210416  BO0REPHA6FP7C15B0K9N1GVO0050               360       iOS   

       PlayEventType                           VideoTitle  
0  

In [70]:
# counts the number of NULLS
df.isnull().sum()

DateID              0
CustomerID          0
TotalTimeWatched    0
Platform            0
PlayEventType       0
VideoTitle          0
dtype: int64

**Observations**
The dataset doesn't contain any null values, thats why it returned 0 on each column

---



In [71]:
#Used to retrieve concise/ brief summary of a dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118534 entries, 0 to 118533
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   DateID            118534 non-null  int64 
 1   CustomerID        118534 non-null  object
 2   TotalTimeWatched  118534 non-null  int64 
 3   Platform          118534 non-null  object
 4   PlayEventType     118534 non-null  object
 5   VideoTitle        118534 non-null  object
dtypes: int64(2), object(4)
memory usage: 5.4+ MB


**Observations**

*  Output quickly reveals the number of rows, columns, data types, and presence of missing values in each column, along with the memory footprint of the Dataframe.



In [72]:
# displays the dataset
print(Viewers_Analysis)

          DateID                    CustomerID  TotalTimeWatched  Platform  \
0       20201101  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
1       20201101  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
2       20201101  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
3       20201101  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
4       20201101  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
...          ...                           ...               ...       ...   
118529  20210416  DGXM5TJ18W1LANOLKJJN5MFT000G               600  Leanback   
118530  20210416  DMMVDBZFDT1F7WL4BGWQ5KOJ000G              5194       Web   
118531  20210416  1JKBIAY7MEUH6BWX0SCU93GJ000I             47660       Web   
118532  20210416  1ADX18R3B33REZ2T0NK911E4002W              1335  Leanback   
118533  20210416  BO0REPHA6FP7C15B0K9N1GVO0050               360       iOS   

       PlayEventType                           VideoTitle  
0  

In [73]:
# used to display dataset
print(df)

          DateID                    CustomerID  TotalTimeWatched  Platform  \
0       20201101  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
1       20201101  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
2       20201101  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
3       20201101  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
4       20201101  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
...          ...                           ...               ...       ...   
118529  20210416  DGXM5TJ18W1LANOLKJJN5MFT000G               600  Leanback   
118530  20210416  DMMVDBZFDT1F7WL4BGWQ5KOJ000G              5194       Web   
118531  20210416  1JKBIAY7MEUH6BWX0SCU93GJ000I             47660       Web   
118532  20210416  1ADX18R3B33REZ2T0NK911E4002W              1335  Leanback   
118533  20210416  BO0REPHA6FP7C15B0K9N1GVO0050               360       iOS   

       PlayEventType                           VideoTitle  
0  

In [74]:
# counting the total duplicate rows
df.duplicated().sum()

np.int64(9873)

**Observations**

*   Used to count the total of duplicate rows in a dataframe
*   In this dataframe there is 9873 duplicates.




In [75]:
# tells us the length of a string
print(len(Viewers_Analysis))


118534


**Observations**
* the length of the dataset has 118534



In [76]:
# Convert DateID to Date format
df['DateID']= pd.to_datetime(df['DateID'],format='%Y%m%d')
display (df)

DateID                    CustomerID  TotalTimeWatched  Platform  \
0      2020-11-01  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
1      2020-11-01  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
2      2020-11-01  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
3      2020-11-01  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
4      2020-11-01  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
...           ...                           ...               ...       ...   
118529 2021-04-16  DGXM5TJ18W1LANOLKJJN5MFT000G               600  Leanback   
118530 2021-04-16  DMMVDBZFDT1F7WL4BGWQ5KOJ000G              5194       Web   
118531 2021-04-16  1JKBIAY7MEUH6BWX0SCU93GJ000I             47660       Web   
118532 2021-04-16  1ADX18R3B33REZ2T0NK911E4002W              1335  Leanback   
118533 2021-04-16  BO0REPHA6FP7C15B0K9N1GVO0050               360       iOS   

       PlayEventType                           VideoTitle  
0             LiveTV            F1 '20: Emilia Romagna GP  
1             LiveTV            F1 '20: Emilia Romagna GP  
2              Other                      Chasing The Sun  
3              Other                      Chasing The Sun  
4             LiveTV                   Sonic The Hedgehog  
...              ...                                  ...  
118529        LiveTV                              New Day  
118530        LiveTV  Buying And Selling With Property...  
118531        LiveTV                             Expresso  
118532        LiveTV                              Unknown  
118533        LiveTV                               Imbewu  

[118534 rows x 6 columns]

**Observations**
*   DateID converted to date format YYYY-MM-DD



In [77]:
# ✅ 6. Saving the updated data
df.to_csv("updated_data.csv", index=False)
display(df)

DateID                    CustomerID  TotalTimeWatched  Platform  \
0      2020-11-01  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
1      2020-11-01  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
2      2020-11-01  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
3      2020-11-01  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
4      2020-11-01  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
...           ...                           ...               ...       ...   
118529 2021-04-16  DGXM5TJ18W1LANOLKJJN5MFT000G               600  Leanback   
118530 2021-04-16  DMMVDBZFDT1F7WL4BGWQ5KOJ000G              5194       Web   
118531 2021-04-16  1JKBIAY7MEUH6BWX0SCU93GJ000I             47660       Web   
118532 2021-04-16  1ADX18R3B33REZ2T0NK911E4002W              1335  Leanback   
118533 2021-04-16  BO0REPHA6FP7C15B0K9N1GVO0050               360       iOS   

       PlayEventType                           VideoTitle  
0             LiveTV            F1 '20: Emilia Romagna GP  
1             LiveTV            F1 '20: Emilia Romagna GP  
2              Other                      Chasing The Sun  
3              Other                      Chasing The Sun  
4             LiveTV                   Sonic The Hedgehog  
...              ...                                  ...  
118529        LiveTV                              New Day  
118530        LiveTV  Buying And Selling With Property...  
118531        LiveTV                             Expresso  
118532        LiveTV                              Unknown  
118533        LiveTV                               Imbewu  

[118534 rows x 6 columns]

**Observations**

*  Saved the updated dataset, with the changed date format





In [78]:
# Counting the different platforms used
platform_counts = df["Platform"].value_counts()
print(platform_counts)


Platform
Leanback    77816
Web         22550
Android     13197
iOS          4971
Name: count, dtype: int64


**Observations**
*  Counting each platform used in the dataset



In [79]:
# Remove exact duplicate rows
df_cleaned = df.drop_duplicates()
print(df_cleaned)

           DateID                    CustomerID  TotalTimeWatched  Platform  \
0      2020-11-01  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
2      2020-11-01  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
4      2020-11-01  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
5      2020-11-01  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
6      2020-11-01  6PMV67PLJ2S47S68J0Y30XFK003C               120       iOS   
...           ...                           ...               ...       ...   
118529 2021-04-16  DGXM5TJ18W1LANOLKJJN5MFT000G               600  Leanback   
118530 2021-04-16  DMMVDBZFDT1F7WL4BGWQ5KOJ000G              5194       Web   
118531 2021-04-16  1JKBIAY7MEUH6BWX0SCU93GJ000I             47660       Web   
118532 2021-04-16  1ADX18R3B33REZ2T0NK911E4002W              1335  Leanback   
118533 2021-04-16  BO0REPHA6FP7C15B0K9N1GVO0050               360       iOS   

       PlayEventType                           Vide

**Observations**
*  Removed rows that are an exact duplicate to others



In [80]:
 # Saving a new CSV file with no duplicated rows
df_cleaned.to_csv('cleaned_output.csv', index=False)
display (df_cleaned)

DateID                    CustomerID  TotalTimeWatched  Platform  \
0      2020-11-01  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
2      2020-11-01  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
4      2020-11-01  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
5      2020-11-01  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
6      2020-11-01  6PMV67PLJ2S47S68J0Y30XFK003C               120       iOS   
...           ...                           ...               ...       ...   
118529 2021-04-16  DGXM5TJ18W1LANOLKJJN5MFT000G               600  Leanback   
118530 2021-04-16  DMMVDBZFDT1F7WL4BGWQ5KOJ000G              5194       Web   
118531 2021-04-16  1JKBIAY7MEUH6BWX0SCU93GJ000I             47660       Web   
118532 2021-04-16  1ADX18R3B33REZ2T0NK911E4002W              1335  Leanback   
118533 2021-04-16  BO0REPHA6FP7C15B0K9N1GVO0050               360       iOS   

       PlayEventType                           VideoTitle  
0             LiveTV            F1 '20: Emilia Romagna GP  
2              Other                      Chasing The Sun  
4             LiveTV                   Sonic The Hedgehog  
5             LiveTV                            Frozen II  
6             LiveTV          PL 20/21: Man Utd v Arsenal  
...              ...                                  ...  
118529        LiveTV                              New Day  
118530        LiveTV  Buying And Selling With Property...  
118531        LiveTV                             Expresso  
118532        LiveTV                              Unknown  
118533        LiveTV                               Imbewu  

[108661 rows x 6 columns]

In [81]:
#Counting the different platforms used in the updated dataset
print(df_cleaned['Platform'].value_counts())
display(df_cleaned)

Platform
Leanback    71525
Web         20733
Android     11582
iOS          4821
Name: count, dtype: int64


DateID                    CustomerID  TotalTimeWatched  Platform  \
0      2020-11-01  EW1DENH0EC1J3M9WAOZF9LSV004O               300  Leanback   
2      2020-11-01  6TS2LLY0L3G66FVY86Q0JEZE000K               360  Leanback   
4      2020-11-01  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
5      2020-11-01  6PMV67PLJ2S47S68J0Y30XFK003C               120  Leanback   
6      2020-11-01  6PMV67PLJ2S47S68J0Y30XFK003C               120       iOS   
...           ...                           ...               ...       ...   
118529 2021-04-16  DGXM5TJ18W1LANOLKJJN5MFT000G               600  Leanback   
118530 2021-04-16  DMMVDBZFDT1F7WL4BGWQ5KOJ000G              5194       Web   
118531 2021-04-16  1JKBIAY7MEUH6BWX0SCU93GJ000I             47660       Web   
118532 2021-04-16  1ADX18R3B33REZ2T0NK911E4002W              1335  Leanback   
118533 2021-04-16  BO0REPHA6FP7C15B0K9N1GVO0050               360       iOS   

       PlayEventType                           VideoTitle  
0             LiveTV            F1 '20: Emilia Romagna GP  
2              Other                      Chasing The Sun  
4             LiveTV                   Sonic The Hedgehog  
5             LiveTV                            Frozen II  
6             LiveTV          PL 20/21: Man Utd v Arsenal  
...              ...                                  ...  
118529        LiveTV                              New Day  
118530        LiveTV  Buying And Selling With Property...  
118531        LiveTV                             Expresso  
118532        LiveTV                              Unknown  
118533        LiveTV                               Imbewu  

[108661 rows x 6 columns]

**Observations**

*   Noted that the count of the different Platforms used changed after removing duplicates

*  **Leanback** with difference of 6291 # meaning there was 6291 duplicated rows
* **Web** difference of 1817 # meaning 1817 duplicated rows
*  **Android** difference 1615 # meaning 1615 duplicated rows
*  **iOS** difference  of 150 # meaning 150 duplicated rows







In [ ]:
print(df_cleaned.groupby('CustomerID')['TotalTimeWatched_Hours'].sum().sort_values(ascending=False))
display(df_cleaned)

**Observations**


*  Total watched time per user
*   CustomerID 18OX855ZBCKS6I1C7X7EISRE001K spent a lot of time watching   







In [85]:
# Most watched titles
print(df_cleaned['VideoTitle'].value_counts().head(10))


VideoTitle
Unknown                      33892
South Africa Tonight          1151
Suidooster                    1145
The Block Australia            890
The South African Morning      744
7de Laan                       711
Binnelanders                   679
The Queen                      666
Love Island South Africa       635
Arendsvlei                     598
Name: count, dtype: int64


**Observations**


*   Top three watched VideoTitle Unknown, South Africa Tonight and Suidooster


In [87]:
# Count unique customers per platform
platform_customer_count = df_cleaned.groupby('Platform')['CustomerID'].nunique().reset_index()
# Rename columns
platform_customer_count.columns = ['Platform', 'CustomerCount']
# Display result
print(platform_customer_count)

   Platform  CustomerCount
0   Android            440
1  Leanback            580
2       Web            398
3       iOS            194


**Observations**

*   Most customers are in the Leanback platform with 580 customer count, 440 on the Android and 398 on the Web platform




In [88]:
# Compare engagement between LiveTV and Catch Up.
print(df_cleaned.groupby('PlayEventType')['TotalTimeWatched_Hours'].sum())


PlayEventType
Catch Up      6906.175556
Downloads       74.910278
LiveTV       52300.841111
Other         5312.304167
Name: TotalTimeWatched_Hours, dtype: float64


**Observations**
*   Viewers engagement greater with Catch Up



In [ ]:
 # Classifying content by title keyword
def classify_genre(title):
    title = title.lower()
    if 'f1' in title or 'ucl' in title or 'pl' in title:
        return 'Sports'
    elif 'amazing hotels' in title or 'documentary' in title:
        return 'Documentary'
    elif 'peppa pig' in title or 'cartoon' in title:
        return 'Kids'
    elif 'murder' in title or 'crime' in title:
        return 'Crime'
    else:
        return 'Other'
df_cleaned['Genre'] = df_cleaned['VideoTitle'].apply(classify_genre)
# Saving updated dataset
df_cleaned.to_csv("enriched_dataset.csv", index=False)
print(df_cleaned['Genre'].value_counts())

**Observations**


* Categorized the shows as per genre, e.g peppa pig classed as cartoon F1 as Sports



In [90]:
# count number of view events
views_by_day = df_cleaned.groupby('DayOfWeek')['TotalTimeWatched_Hours'].count().sort_values(ascending=False)
print(views_by_day)

DayOfWeek
Sunday       17477
Saturday     16383
Thursday     15419
Tuesday      14996
Friday       14978
Wednesday    14946
Monday       14462
Name: TotalTimeWatched_Hours, dtype: int64


**Observations**

*   Number of views each day of the week



In [91]:
#  Which videos were watched more than once (might be because of repeats).
df_cleaned['VideoTitle'].value_counts()
print(df_cleaned['VideoTitle'].value_counts(5))

VideoTitle
Unknown                            0.311906
South Africa Tonight               0.010593
Suidooster                         0.010537
The Block Australia                0.008191
The South African Morning          0.006847
                                     ...   
BBC World News America             0.000009
the eighties                       0.000009
Zoom In: #HBD Chance The Rapper    0.000009
Austrian Open                      0.000009
Zoom In: Pre-Party                 0.000009
Name: proportion, Length: 8904, dtype: float64


**Observations**


*  Videos or shows which were watched more than once



In [92]:
# which videos were watched longer on average per event.
df_cleaned.groupby('VideoTitle')['TotalTimeWatched_Hours'].mean().sort_values(ascending=False)
print(df_cleaned.groupby('VideoTitle')['TotalTimeWatched_Hours'].mean().sort_values(ascending=False))


VideoTitle
BOX HL '21: Alvarez v Yildirim         18.160833
Free In Deed                           14.600000
ploey                                  12.800000
WGC HL '21 WGC at The Concession D3    10.303333
Black Market                            9.533333
                                         ...    
Closet Confidential                     0.000278
Prophet Sbu Mokoena                     0.000278
Jesus Trejo: Stay At Home Son           0.000278
Until Forever                           0.000278
Besigheid wat saak maak                 0.000278
Name: TotalTimeWatched_Hours, Length: 8904, dtype: float64


**Observations**

VideoTitles with the highest average time
*   BOX HL '21: Alvarez v Yildirim
*   Free In Deed
*   ploey




In [ ]:
# After class 14/07/25 asked to create time buckets showing times each viewer spent

# Create a function to convert seconds in the dataset to hh:mm:ss
def convert_to_hms(seconds):
    return time.strftime("%H:%M:%S", time.gmtime(seconds))
# Apply the function to the 'TotalTimeWatched' column
df_cleaned['TimeWatched_hh:mm:ss'] = df_cleaned['TotalTimeWatched'].apply(convert_to_hms)
# Display result
print(df_cleaned[['TotalTimeWatched', 'TimeWatched_hh:mm:ss']].tail().sort_values(by='TotalTimeWatched',ascending=False))
# Save to new CSV
df_cleaned.to_csv('viewership_with_hms.csv', index=False)
display(df_cleaned)



**Observations**
* Converted the TotalTimeWatched to hh:mm:ss format, so it will be readable

In [111]:
 # Creating watch time buckets, whose watching the most and whose watching less
 # Convert TotalTimeWatched (in seconds) to timedelta
df['WatchDuration'] = pd.to_timedelta(df['TotalTimeWatched'], unit='s')
# Define the classification function using timedelta
def classify_viewer(duration):
    if pd.Timedelta('0:00:00') <= duration <= pd.Timedelta('1:30:00'):
        return 'Selective_Viewer'
    elif pd.Timedelta('1:30:01') <= duration <= pd.Timedelta('4:30:00'):
        return 'Average_Viewer'
    elif pd.Timedelta('4:30:01') <= duration <= pd.Timedelta('6:30:00'):
        return 'Enthusiast_Viewer'
    else:
        return 'Other'
# Apply to the new timedelta column
df['ViewerType'] = df['WatchDuration'].apply(classify_viewer)
print(df['ViewerType'].value_counts())




ViewerType
Selective_Viewer     108793
Average_Viewer         8499
Enthusiast_Viewer       674
Other                   568
Name: count, dtype: int64


**Observations**


*  ViewerType based on time sent watching each show
*  ViewerType
   Selective_Viewer     99152
   Average_Viewer        8309
   Enthusiast_Viewer      665
   Other                  535

